File: src/types/service.ts
import { Order } from './order';

export interface IServiceCategory {
   id: string;
   name: string;
   description?: string;
   is_active: boolean;
   created_at: string;
   updated_at: string;
}

export interface IService {
   id: string;
   service_category_id?: string;
   service_category?: IServiceCategory;
   parent_id?: string | null;    // null = top-level service, set = variant of parent
   parent?: IService;            // populated when is variant
   variants?: IService[];        // sub-services (variants)
   name: string;
   description?: string;
   price: number;
   unit: string; // 'kg', 'pcs', 'set'
   estimated_duration?: number; // in minutes
   estimate_hour?: number;      // in hours
   minimum_qty_order?: number;  // BOM set size, default 1
   is_active: boolean;
   is_parent?: boolean;
   images?: string; // JSON string from backend
   service_cogs?: IServiceCog[];
   created_at: string;
   updated_at: string;
}

export interface IServiceCog {
   id: string;
   service_id: string;
   raw_material_id: string;
   raw_material?: {
      id: string;
      title: string;
      unit: string;
      current_stock: number;
   };
   qty: number;
   unit: string;
}

export interface IOrderService {
   id: string;
   order_id: string;
   service_id: string;
   service?: IService;
   service_variant_id?: string | null;
   service_variant?: IService;
   qty: number;
   price_at_order: number;
   subtotal: number;
   notes?: string;

   order?: Order;

   // Tracking
   service_process?: IOrderServiceProcess[];
   service_detail?: IOrderServiceDetail;
}


export interface IOrderServiceProcess {
   id: string;
   order_service_id: string;
   process_type: string; // pickup, processing, delivery, done, other
   description?: string;
   images?: string; // JSON string
   created_by?: string;
   creator?: {
      first_name: string;
      last_name: string;
      email: string;
   };
   created_at: string;
}

export interface IOrderServiceDetail {
   id: string;
   order_service_id: string;
   description?: string; // HTML
   images?: string; // JSON string
   created_at: string;
   updated_at: string;
}

export type IServiceTableFilterValue = string | string[];

export type IServiceTableFilters = {
   name: string;
   status: string;
};

File: modules/cms/models/service.go
package models

import (
	"aldev/modules/global/models"

	"github.com/google/uuid"
)

type Service struct {
	models.BaseModel
	ServiceCategoryID *uuid.UUID       `json:"service_category_id" gorm:"type:uuid"`
	ServiceCategory   *ServiceCategory `json:"service_category,omitempty" gorm:"foreignKey:ServiceCategoryID"`
	ParentID          *uuid.UUID       `json:"parent_id,omitempty" gorm:"type:uuid;index"` // null = parent service, set = variant of parent
	Parent            *Service         `json:"parent,omitempty" gorm:"foreignKey:ParentID"`
	Variants          []Service        `json:"variants,omitempty" gorm:"foreignKey:ParentID"`
	Name              *string          `json:"name" gorm:"type:varchar(200);not null" validate:"required,min=2,max=200"`
	Description       *string          `json:"description,omitempty" gorm:"type:text"`
	Price             *float64         `json:"price" gorm:"type:decimal(15,2);not null" validate:"required,gte=0"`
	Unit              *string          `json:"unit" gorm:"type:varchar(20);not null" validate:"required"`
	EstimatedDuration *int             `json:"estimated_duration,omitempty"`                          // in minutes
	EstimateHour      *float64         `json:"estimate_hour,omitempty" gorm:"type:decimal(10,2)"`     // in hours
	MinimumQtyOrder   *float64         `json:"minimum_qty_order" gorm:"type:decimal(10,2);default:1" validate:"omitempty,min=1"` // BOM set size, default 1
	IsActive          *bool            `json:"is_active" gorm:"default:true"`
	IsParent          *bool            `json:"is_parent" gorm:"default:false"`
	Images            *string          `json:"images,omitempty" gorm:"type:text"` // JSON array
	ServiceCogs       []ServiceCog     `json:"service_cogs,omitempty" gorm:"foreignKey:ServiceID"`
}
